design/metadata-handling.md

```diff
@@ -0,0 +1,750 @@
+spec:
+  metaData:
+    - key: abc
+      string: def
+    - key: name_m3m
+      objectName: metal3machine
+    - key: name_machine
+      objectName: machine
+    - key: name_bmh
+      objectName: baremetalhost
+    - key: index
+      index:
+        offset: 0
+        step: 1
+    - key: ip
+      ip:
+        start: 192.168.0.10
+        end: 192.168.0.100
+        subnet: 192.168.0.0/24
+        step: 1
+    - key: mac
+      fromHostInterface: "eth0"
+  networkData:
+    links:
+      - ethernet:
+          type: "phy"
+          id: "enp1s0"
+          mtu: 1500
+          macAddress:
+            fromHostInterface: "eth0"
+      - ethernet:
+          type: "phy"
+          id: "enp2s0"
+          mtu: 1500
+          macAddress:
+            fromHostInterface: "eth1"
+      - bond:
+          id: "bond0"
+          mtu: 1500
+          macAddress:
+            string: "XX:XX:XX:XX:XX:XX"
+          bondMode: "802.1ad"
+          bondLinks:
+            - enp1s0
+            - enp2s0
+      - vlan:
+          id: "vlan1"
+          mtu: 1500
+          macAddress:
+            string: "YY:YY:YY:YY:YY:YY"
+          vlanId: 1
+          vlanLink: bond0
+    networks:
+      - ipv4DHCP:
+          id: "provisioning"
+          link: "bond0"
+
+      - ipv4:
+          id: "Baremetal"
+          link: "vlan1"
+          ipAddress:
+            start: "192.168.0.10"
+            end: "192.168.0.100"
+            subnet: "192.168.0.0/24"
+            step: 1
+          netmask: 24
+          routes:
+            - network: "0.0.0.0"
+              netmask: 0
+              gateway: "192.168.0.1"
+              services:
+                - type: "dns"
+                  address: "8.8.4.4"
+      - ipv6DHCP:
+          id: "provisioning6"
+          link: "bond0"
+      - ipv6SLAAC:
+          id: "provisioning6slaac"
+          link: "bond0"
+      - ipv6:
+          id: "Baremetal6"
+          link: "vlan1"
+          ipAddress:
+            start: "2001:0db8:85a3::8a2e:0370:a"
+            end: "2001:0db8:85a3::8a2e:0370:fff0"
+            subnet: "2001:0db8:85a3::8a2e:0370:0/64"
+            step: 10
+          netmask: 64
+          routes:
+            - network: "0::0"
+              netmask: 0
+              gateway: "2001:0db8:85a3::8a2e:0370:1"
+              services:
+                - dns: "2001:4860:4860::8844"
+    services:
+      - dns: "8.8.8.8"
+      - dns: "2001:4860:4860::8888"
+status:
+  indexes:
+    "machine-1": 0
+  lastUpdated: "2020-04-02T06:36:09Z"
+```
+
+This object will be reconciled by its own controller. When reconciled,
+the controller will add a label pointing to the Metal3Cluster that has nodes
+linking to this object. The spec contains a `metaData` and a `networkData` field
+that contain a template of the values that will be rendered for all nodes.
+
+The `metaData` field will be rendered into a map of strings in yaml format,
+while `networkData` will be rendered into a json string that fulfills the
```

Comment:
gophercloud requires map[string]interface{} for network data

Reply:
fixed, thank you